packages/app/src/cli/services/function/ui/components/hooks/extension-watcher.ts

```diff
@@ -0,0 +1,193 @@
+import {FunctionRunData} from '../../../replay.js'
+import {AppInterface} from '../../../../../models/app/app.js'
+import {FunctionConfigType} from '../../../../../models/extensions/specifications/function.js'
+import {ExtensionInstance} from '../../../../../models/extensions/extension-instance.js'
+import {setupExtensionWatcher} from '../../../../dev/extension/bundler.js'
+import {exec} from '@shopify/cli-kit/node/system'
+import {AbortController} from '@shopify/cli-kit/node/abort'
+import {useEffect, useRef, useState} from 'react'
+import {useInput, useStdin} from '@shopify/cli-kit/node/ink'
+import {handleCtrlC, renderFatalError} from '@shopify/cli-kit/node/ui'
+import {useAbortSignal} from '@shopify/cli-kit/node/ui/hooks'
+import {isUnitTest} from '@shopify/cli-kit/node/context/local'
+import {treeKill} from '@shopify/cli-kit/node/tree-kill'
+import {FatalError} from '@shopify/cli-kit/node/error'
+import {outputWarn} from '@shopify/cli-kit/node/output'
+import {Writable} from 'stream'
+
+interface FunctionRun {
+  type: 'functionRun'
+  input: string
+  output: string
+  logs: string
+  name: string
+  size: number
+  memory_usage: number
+  instructions: number
+}
+
+interface SystemMessage {
+  type: 'systemMessage'
+  message: string
+}
+
+type ReplayLog = FunctionRun | SystemMessage
+
+interface WatchFunctionForReplayOptions {
+  selectedRun: FunctionRunData
+  abortController: AbortController
+  app: AppInterface
+  extension: ExtensionInstance<FunctionConfigType>
+}
+
+export function setupExtensionWatcherForReplay({
```

Comment:
Nit- I think `setupExtensionWatcherForReplay` would be a more appropriate name for this file since it's not a generic utility for watching functions, but a specific function for replay. That brings it in line with the app logs precedent of `usePollAppLogs.ts`